packages/@headlessui-react/src/components/dialog/dialog.tsx

```diff
@@ -212,6 +216,18 @@ let InternalDialog = forwardRefWithAs(function InternalDialog<
     ]),
   })
 
+  // Ensure that the Dialog is the top layer when it is opened.
+  //
+  // In a perfect world this is pushed / popped when we open / close the Dialog
+  // for within an event listener. But since the state is controlled by the
+  // user, this is the next best thing to do.
+  let stackMachine = stackMachines.get(null)
```

Comment:
I believe this should be safe because we are using a `DefaultMap` in module scope and this will always return the exact same reference.